src/codegen/array_boundary.rs

```diff
@@ -0,0 +1,159 @@
+                    BigInt::one(),
+                )),
+            )
+        } else {
+            Expression::Add(
+                Loc::Codegen,
+                Type::Uint(32),
+                false,
+                Box::new(Expression::Variable(Loc::Codegen, Type::Uint(32), *to_add)),
+                Box::new(Expression::NumberLiteral(
+                    Loc::Codegen,
+                    Type::Uint(32),
+                    BigInt::one(),
+                )),
+            )
+        };
+
+        // Add instruction to the cfg
+        cfg.add(
+            vartab,
+            Instr::Set {
+                loc: Loc::Codegen,
+                res: *to_add,
+                expr: add_expr,
+            },
+        );
+    }
+}
+
+pub(crate) fn handle_array_assign(
+    right: &ast::Expression,
+    cfg: &mut ControlFlowGraph,
+    vartab: &mut Vartable,
+    pos: &usize,
+) {
+    if let ast::Expression::Variable(_, _, right_res) = *right {
+        // If we have initialized a temp var for this var
+        if let Some(to_update) = cfg.array_lengths_temps.clone().get(&right_res) {
+            let num = Expression::Variable(Loc::Codegen, Type::Uint(32), *to_update);
+            let temp_res = vartab.temp_name("array_size", &Type::Uint(32));
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc: Loc::Codegen,
+                    res: temp_res,
+                    expr: num,
+                },
+            );
+
+            cfg.array_lengths_temps.insert(*pos, temp_res);
+        } else {
+            // If the right hand side doesn't have a temp, it must be a function parameter.
+            cfg.array_lengths_temps.remove(pos);
+        }
+    } else if let ast::Expression::AllocDynamicArray(_, ty, size, _) = right {
+        let a = if cfg.array_lengths_temps.get(pos) != None {
+            // If there was a temp variable, reassign it to the new value
+            cfg.array_lengths_temps.remove(pos).unwrap()
+        } else {
+            //create a new temp variable
+            vartab.temp_name("array_size", ty)
+        };
+
+        if let ast::Expression::Variable(_, _, position) = **size {
+            let num = Expression::Variable(Loc::Codegen, Type::Uint(32), position);
+            cfg.add(
+                vartab,
+                Instr::Set {
+                    loc: Loc::Codegen,
+                    res: a,
+                    expr: num,
+                },
+            );
+
+            cfg.array_lengths_temps.insert(*pos, a);
+        }
+        // If size a uint and bits > 32
+        else if let ast::Expression::Trunc(_, _, ref index) = **size {
+            if let ast::Expression::Variable(_, _, index) = &**index {
+                // A number var holding array length
+                let num = Expression::Variable(Loc::Codegen, Type::Uint(32), *index);
+                let num_trunced = Expression::Trunc(Loc::Codegen, Type::Uint(32), Box::new(num));
```

Comment:
The same case as above. You already have an `Expression::Trunc`, there is no need to create it again. In addition, this construction is wrong. The variable is not 32-bit wide, so creating `Expression::Variable(Loc::Codegen, Type::Uint(32), *index)` will generate incorrect LLVM code.